### PR TITLE
Randomized string concat

### PR DESCRIPTION
**Pull request by alekseyl**
proposition for https://github.com/fastruby/fast-ruby/issues/158 fix.

### CODE REVIEW

**Review by JuanVqz — COMMENTED** [review]
Hey, @alekseyl thank you for the time you spent on this, it looks good, I just left some minor comments, let me know what you think :+1:

**Comment by alekseyl** [comment]
Thnx @JuanVqz, no worries for the delay :). I've commited suggested!

File: code/string/concatenation_randomized.rb [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+      sarr2 = Array.new(2_000_000) { RandStr.eq100.dup }
+  
+      i, j = 0, 0
+      x.report("Collateral actions only")  { k=0; 1_000_000.times { k+=1; RandStr.eq10; sarr2[k]; RandStr.lt100; } }
+      x.report("String#+")         { k=0; 1_000_000.times { k+=1; sarr1[k]; fastest_plus(RandStr.eq10, RandStr.lt100) } }
+      x.report('"#{foo}#{bar}"')   { k=0; 1_000_000.times { k+=1; sarr2[k]; fast_interpolation(RandStr.eq10, RandStr.lt100) } }
+      x.report("String#concat")    { 1_000_000.times { RandStr.eq10; slow_concat(sarr1[i], RandStr.lt100); i+=1; } }
+      x.report("String#append")    { 1_000_000.times { RandStr.eq10; slow_append(sarr2[j], RandStr.lt100); j+=1;  } }
+    end
+end
```
Comment (JuanVqz) [inline code/string/concatenation_randomized.rb:154]:
```suggestion
def bench_10_to_100
  Benchmark.bmbm do |x|
    sarr1 = Array.new(2_000_000) { RandStr.eq100.dup }
    sarr2 = Array.new(2_000_000) { RandStr.eq100.dup }
  
    i, j = 0, 0
    x.report("Collateral actions only")  { k=0; 1_000_000.times { k+=1; RandStr.eq10; sarr2[k]; RandStr.lt100; } }
    x.report("String#+")         { k=0; 1_000_000.times { k+=1; sarr1[k]; fastest_plus(RandStr.eq10, RandStr.lt100) } }
    x.report('"#{foo}#{bar}"')   { k=0; 1_000_000.times { k+=1; sarr2[k]; fast_interpolation(RandStr.eq10, RandStr.lt100) } }
    x.report("String#concat")    { 1_000_000.times { RandStr.eq10; slow_concat(sarr1[i], RandStr.lt100); i+=1; } }
    x.report("String#append")    { 1_000_000.times { RandStr.eq10; slow_append(sarr2[j], RandStr.lt100); j+=1;  } }
  end
end
```

**Review by JuanVqz — APPROVED** [review]
@alekseyl, Thank you for your contribution, it looks good :+1:

**Comment by JuanVqz** [comment]
@mateusdeap could you please have a look to merge it?

**Review by etagwerker — CHANGES_REQUESTED** [review]
@alekseyl @JuanVqz I noticed that this PR uses `bmbm` and we want all benchmarks to use `ips`

Could you please tweak this benchmark to use "benchmark/ips"? 

It shouldn't affect the results. 

Thanks!

**Comment by alekseyl** [comment]
> @alekseyl @JuanVqz I noticed that this PR uses bmbm and we want all benchmarks to use ips
> Could you please tweak this benchmark to use "benchmark/ips"?
> It shouldn't affect the results.
> Thanks!

@etagwerker Hi Ernesto!

Sorry for the delayed answer. The problem is it will not look as good as we migh expect, because of a randomization )). It will show a much bigger variance (20-30-40% I can't recall properly) over relatively "small" difference. 

I might give it a try one more time, but can't promise that it will be soon enough :(. I need an inspiration to dive in to that river again  ))).

**Comment by ixti** [comment]
Can we also add:

``` ruby
[foo, bar].join
```

It would also make sense to check if there will be different results with more than 2 strings to concatenate. Also, would be nice to see performance comparison of cases like:

``` ruby
"#{foo}-#{bar}"
```

**Review by etagwerker — APPROVED** [review]
@alekseyl Looks good, thank you!

**Pull request merged**